src/ScheduleFunctions.cpp

```diff
@@ -2224,9 +2226,51 @@ bool validate_schedule(Function f, const Stmt &s, const Target &target, bool is_
         }
     }
 
-    // Check there isn't a parallel loop between the compute_at and the store_at
     std::ostringstream err;
 
+    // Additional sites to exclude from the allowed sites, based on (eg) GPU constraints
+    set<int> invalid_sites;
+
+    // If you're compute_at() inside a gpu blocks loop, you can't have a gpu blocks loop yourself
+    const auto has_gpu_blocks = [&]() {
+        for (const Dim &d : f.definition().schedule().dims()) {
+            if (d.for_type == ForType::GPUBlock) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    if (store_at_ok && compute_at_ok && has_gpu_blocks()) {
+        for (size_t i = 0; i <= compute_idx; i++) {
+            if (sites[i].is_gpu_block) {
+                string site_fname = sites[i].loop_level.func();
+                user_error << "Functions that are compute_at() a gpu_block() loop cannot have their own gpu_block() loops, "
+                           << "but Func \"" << f.name() << "\" is compute_at() \"" << site_fname << "\"\n";
+            }
+        }
+    }
+
+    // If you're compute_at() a var marked as a gpu block var, it must be the innermost one
+    if (store_at_ok && compute_at_ok && sites[compute_idx].is_gpu_block) {
+        string compute_at_fname = sites[compute_idx].loop_level.func();
+        int possibly_invalid_idx = compute_idx;
+        for (size_t i = compute_idx + 1; i < sites.size(); i++) {
+            if (!sites[i].is_gpu_block) {
+                continue;
+            }
+            string site_fname = sites[i].loop_level.func();
+            if (site_fname == compute_at_fname) {
+                err << "Functions that are compute_at() a gpu_block() loop must specify the innermost gpu_block() loop for that Func.\n";
+                invalid_sites.insert(possibly_invalid_idx);
```

Comment:
I'm confused. Why keep a set of invalid sites, as opposed to just removing them from the set of valid sites?

Reply:
When I first wrote this I thought it might be confusing to the user to just leave out those locations (since they would be valid for cpu code), and thought maybe they should be listed specially for that reason (i.e., "These locations aren't usable with GPU because blah blah")... but maybe I'm overthinking it. I'll simplify it as you suggest.

Reply:
(but there actually was a good reason to make a list of invalid sites: if we prune the invalid sites as we find them, we might invalidate the indices for for `compute_idx` and `store_idx`)